### PR TITLE
Implement `put_report` and test `http_post_upload`

### PR DESCRIPTION
**Pull request by nakatsuka-y**
This implements the `put_report` method called within `http_post_upload`.
`http_post_upload` is used by the Leader to handle http post requests from Clients.

The changes include unit tests that check for malformed reports as well as a test that successfully receives a report.

Ref: #20.

### CODE REVIEW

File: daphne/src/roles_test.rs [hunk 1]
```diff
@@ -367,9 +367,10 @@ async fn http_post_upload_fail_send_invalid_report() {
     );
 }
 
-// This tests a successful Client --> Leader functionality.
+// This currently tests successful Client --> Leader functionality.
```
Comment (cjpatton) [inline daphne/src/roles_test.rs:370]: By the way, it is still useful to have this test in addition to a full end-to-end test. It provides more feedback to the developer in case a change is causing one to fail but not the other.

```suggestion
// Test the upload sub-protocol.
```

File: daphne/src/roles_test.rs [hunk 2]
```diff
@@ -20,6 +23,18 @@ use prio::codec::{Decode, Encode};
 use rand::{thread_rng, Rng};
 use std::{ops::DerefMut, vec};
 
+fn gen_test_upload_req(agg: &MockAggregator, report: Report) -> DapRequest<BearerToken> {
```
Comment (cjpatton) [inline daphne/src/roles_test.rs:26]: Suggestion (here and below): It would be a bit cleaner to define this function as part of MockAggregator's implementation:

```rust
impl MockAggregator {
    fn gen_test_upload(&self, report: Report) -> DapRequest<BearerToken> {
        ...
    }
}
```

File: daphne/src/roles_test.rs [hunk 3]
```diff
@@ -299,3 +314,69 @@ async fn http_post_aggregate_abort_helper_state_overwritten() {
         assert_eq!(e, "unexpected message for aggregation job (already exists)")
     );
 }
+
+#[tokio::test]
+async fn http_post_upload_fail_send_invalid_report() {
+    let leader = MockAggregator::new();
+    let task_id = leader.nominal_task_id();
+
+    // Construct a report payload with a 0'd out task ID.
```
Comment (cjpatton) [inline daphne/src/roles_test.rs:323]: It doesn't matter that it's 0'd out; what is relevant is that it's not the right task ID.

File: daphne/src/lib.rs [hunk 4]
```diff
@@ -95,6 +95,12 @@ impl From<::hpke::HpkeError> for DapError {
     }
 }
 
+impl From<TransitionFailure> for DapError {
```
Comment (cjpatton) [inline daphne/src/lib.rs:98]: I think the code would be clearer without this. Beware of overusing the `From<A> for B` -- in general it's best to handle things explicitly.

**Review by cjpatton — CHANGES_REQUESTED** [review]
Nice, just a few comments. Also: In the future, before writing tests, check the rest of the crate to see if there is already test code that covers the change. Adding more tests than needed can't hurt, but it does add maintenance burden, so it's important to think carefully about what we need. In particular, `vdaf/mod_test.rs` already has tests that exercise some of the corner cases in the new tests.

That said, let's keep the new tests you added. Just something to keep in mind for the future :)

File: daphne/src/roles_test.rs (hunk 1 above)
Comment (nakatsuka-y) [reply daphne/src/roles_test.rs]: I've added the sub-protocol test as well.

File: daphne/src/roles_test.rs (hunk 2 above)
Comment (nakatsuka-y) [reply daphne/src/roles_test.rs]: Great suggestion, I think this would make the code a bit more readable as well.

File: daphne/src/roles_test.rs (hunk 3 above)
Comment (nakatsuka-y) [reply daphne/src/roles_test.rs]: Right. Updated comment.

File: daphne/src/lib.rs (hunk 4 above)
Comment (nakatsuka-y) [reply daphne/src/lib.rs]: I thought the code would look cleaner because we just have to use `?`, but I have removed this and reverted back to handling the error explicitly.

**Review by cjpatton — COMMENTED** [review]
Changes look good, go ahead and squash. By the way the title of of this PR is "Implement and test http_post_upload", which is not entirely accurate. What is implemented is `put_report` for `MockAggregator`, which allows `http_post_upload` to be tested.

**Review by cjpatton — APPROVED** [review]
